browser/ui/tabs/brave_tab_strip_model.cc

```diff
@@ -0,0 +1,146 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <algorithm>
+#include <memory>
+
+#include "brave/browser/ui/tabs/brave_tab_strip_model.h"
+
+#include "brave/common/pref_names.h"
+#include "build/build_config.h"
+#include "chrome/browser/profiles/profile.h"
+#include "components/prefs/pref_service.h"
+#include "content/public/browser/web_contents.h"
+#include "ui/events/event.h"
+#include "ui/events/event_handler.h"
+
+#if defined(USE_AURA)
+#include "ui/aura/window.h"
+#endif
+
+#if defined(OS_MACOSX)
+#include "ui/views/widget/widget.h"
+#endif
+
+BraveTabStripModel::BraveTabStripModel(TabStripModelDelegate* delegate,
+                                       Profile* profile)
+    : TabStripModel(delegate, profile) {}
+
+BraveTabStripModel::~BraveTabStripModel() {}
+
+void BraveTabStripModel::SelectNextTab(UserGestureDetails detail) {
+  bool isMRUEnabled =
+      Profile::FromBrowserContext(GetActiveWebContents()->GetBrowserContext())
+          ->GetPrefs()
+          ->GetBoolean(kMRUCyclingEnabled);
+
+  if (isMRUEnabled) {
+    SelectTabMRU(false, detail);
+  } else {
+    SelectRelativeTab(true, detail);
+  }
+}
+
+void BraveTabStripModel::SelectPreviousTab(UserGestureDetails detail) {
+  bool isMRUEnabled =
+      Profile::FromBrowserContext(GetActiveWebContents()->GetBrowserContext())
+          ->GetPrefs()
+          ->GetBoolean(kMRUCyclingEnabled);
+
+  if (isMRUEnabled) {
+    SelectTabMRU(true, detail);
+  } else {
+    SelectRelativeTab(false, detail);
+  }
+}
+
+void BraveTabStripModel::SelectTabMRU(bool backward,
+                                      UserGestureDetails detail) {
+  if (current_mru_cycling_index == -1) {
+    // Start cycling
+
+    // Create a list of tab indexes sorted by time of last activation
+    for (int i = 0; i < count(); ++i) {
+      mru_cycle_list.push_back(i);
+    }
+
+    std::sort(mru_cycle_list.begin(), mru_cycle_list.end(),
+              [this](int a, int b) {
+                return GetWebContentsAt(a)->GetLastActiveTime() >
+                       GetWebContentsAt(b)->GetLastActiveTime();
+              });
+
+    current_mru_cycling_index = 0;
+
+    // Create an event handler eating all keyboard events while tabing
+    ctrl_released_event_handler = std::make_unique<CtrlReleaseHandler>(this);
+
+    // Add the event handler
+    gfx::NativeWindow window =
+        this->GetActiveWebContents()->GetTopLevelNativeWindow();
+#if defined(OS_MACOSX)
+    views::Widget::GetWidgetForNativeWindow(window)
+        ->GetRootView()
+        ->AddPreTargetHandler(ctrl_released_event_handler.get());
+#else
+    window->AddPreTargetHandler(ctrl_released_event_handler.get());
+#endif
+  }
+
+  int tabCount = mru_cycle_list.size();
+
+  if (backward) {
+    current_mru_cycling_index =
+        (current_mru_cycling_index - 1 % tabCount + tabCount) % tabCount;
```

Comment:
It doesn't work because in C++ (and JS btw) `%` is the reminder operator. So `-1 % 5 = -1`.  In Python for example, you actually get the real modulo operator `-1 % 5 = 4` confusing stuff when you switch language all the time x)